tests/TestCase/View/CellTest.php

```diff
@@ -460,4 +460,24 @@ public function testACachedViewCellReRendersWhenGivenADifferentTemplate(): void
         Cache::delete('celltest');
         Cache::drop('default');
     }
+
+    /**
+     * Tests events are dispatched correctly
+     */
+    public function testCellRenderDispatchesEvents(): void
+    {
+        /** @var \TestApp\View\Cell\ArticlesCell $cell */
+        $cell = $this->View->cell('Articles::doEcho', ['msg1' => 'dummy', 'msg2' => ' message']);
+
+        $manager = $this->View->getEventManager();
+        $manager->on('Cell.beforeAction', function ($event, $eventCell, $action) use ($cell) {
+            $this->assertSame($eventCell, $cell);
+            $this->assertEquals('doEcho', $action);
+        });
+        $manager->on('Cell.beforeAction', function ($event, $eventCell, $action) use ($cell) {
+            $this->assertSame($eventCell, $cell);
+            $this->assertEquals('doEcho', $action);
+        });
+        $this->assertStringContainsString('dummy message', $cell->render());
```

Comment:
This assertion is unrelated to the events execution, can be changed to just `$cell->render()`.

Also for the asserts you have added there won't be any failure reported even if the events are not triggered. To check that the callbacks are run set a var outside the callback, pass it by ref to the callback and modify inside it and then assert it's value outside the callback.